Safe load of types from failing assemblies in TypeUtils.GetTypes

DixonD-git opened the pull request:
Trying to address the problem I've raised in https://github.com/dotnet/orleans/issues/1517 . If the failing assembly (in my case `Microsoft.CodeAnalysis.Scripting`) is loaded into `AppDomain` (so it is returned in `AppDomain.CurrentDomain.GetAssemblies()`), then the silo is crashing on start.

dnfclas commented:
Hi __@DixonD-git__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        This seems like a small (but important) contribution, so no Contribution License Agreement is required at this point. Real humans will now evaluate your PR.
    </span>

TTYL, DNFBOT;

ReubenBond commented on src/Orleans/CodeGeneration/TypeUtils.cs:
```diff
@@ -449,7 +449,22 @@ public static Type ToReflectionOnlyType(Type type)
+            {
+                definedTypes = assembly.DefinedTypes;
+            }
+            catch (ReflectionTypeLoadException)
```
Are there any other exceptions which we need to catch? Should we catch all exceptions?
We should log these exceptions.

ReubenBond commented on src/Orleans/CodeGeneration/TypeUtils.cs:
```diff
@@ -449,7 +449,22 @@ public static Type ToReflectionOnlyType(Type type)
+                return Enumerable.Empty<Type>();
+            }
+
+            return definedTypes.Where(type => !type.GetTypeInfo().IsNestedPrivate && whereFunc(type));
```
This line could be put inside the `try` block, and that could become a simple `return` with no extra variables.

DixonD-git commented:
@ReubenBond Thanks for comments, I've added logging and changed `ReflectionTypeLoadException` to `Exception`  - I've reused the similar logic from `AssemblyProcessor` (I hope it is ok).

ReubenBond commented:
Great work, @DixonD-git, thank you! Looks Good To Me. When the tests pass, I will merge it.

sergeybykov commented:
Tests passed. I'll merge it, so that I can cherry-pick it for 1.1.3.

Thank you, @DixonD-git!
